README.rst
gcovr
this is my gcovr version.
I add the function converage as issue in gcovr (https://github.com/gcovr/gcovr/pull/362/commits)
but gcovr did not give the offical version by add function coverage
so this is only my change version
if you have some question, tell me.


CHANGELOG.rst
``gcovr`` Release History and Change Log

.. program is needed to resolve option links
.. program::  gcovr

Next Release
------------

Breaking changes:

 - Dropped support for Python 2 and Python 3.5.
   From now on, gcovr will only support Python versions
   that enjoy upstream support.

Improvements and new features:

 - Add function coverage metric (:issue:`325`)
 - Handles spaces in ``gcov`` path. (:issue:`385`)
 - Early fail when output cannot be created. (:issue:`382`)
 - Add :option:`--txt` for text output. (:issue:`387`)
 - Add :option:`--csv` for CSV output. (:issue:`376`)
 - Add :option:`--exclude-lines-by-pattern` to filter out source lines by arbitrary
   regex. (:issue:`356`)
 - Add :option:`--json-summary` to generate a :ref:`JSON Summary <json_summary_output>` report. (:issue:`366`)
 - Add :option:`--coveralls` to generate a :ref:`Coveralls <coveralls_output>` compatible JSON report. (:issue:`328`)
 - Add support for output directories. If the output ends with a ``/`` or ``\`` it is used as a directory. (:issue:`416`)
 - Compare paths case insensitive if file system of working directory is case insensitive. (:issue:`329`)
 - Add wildcard pattern to json :option:`--add-tracefile`. (:issue:`351`)
 - Enable :option:`--filter` and :option:`--exclude` for :ref:`Combining tracefiles <combining_tracefiles>`. (:issue:`373`)
 - Only output 100.0% in text and HTML output if really 100.0%, else use 99.9%. (:issue:`389`)
 - Support relative source location for shadow builds. (:issue:`410`)
 - Incorrect path for header now can still generate html-details reports (:issue:`271`)
 - Change format version in JSON output from number to string and update it to "0.2".  (:issue:`418`, :issue:`463`)
 - Only remove :option:`--root` path at the start of file paths. (:issue:`452`)
 - Fix coverage report for cmake ninja builds with given in-source object-directory. (:issue:`453`)
 - Add issue templates. (:issue:`461`)
 - Add :option:`--exclude-function-lines` to exclude the line of the function definition in the coverage report. (:issue:`430`)  
 - Changes for HTML output format:

   - Redesign HTML generation. Add :option:`--html-self-contained` to control externeal or internal CSS. (:issue:`367`)
   - Change legend for threshold in html report. (:issue:`371`)
   - Use HTML title also for report heading. Default value for :option:`--html-title` changed. (:issue:`378`)
   - Add :option:`--html-tab-size` to configure tab size in HTML details. (:issue:`377`)
   - Add option :option:`--html-css` for user defined styling. (:issue:`380`)
   - Create details html filename independent from OS. (:issue:`375`)
   - Add :option:`--html-theme` to change the color theme. (:issue:`381`)
   - Add linkable lines in HTML details. (:issue:`401`)
   - Add syntax highlighting in the details HTML report. This can be turned off with :option:`--no-html-details-syntax-highlighting <--html-details-syntax-highlighting>`. (:issue:`402`, :issue:`415`)

Documentation:

 - Cookbook: :ref:`oos cmake` (:issue:`340`, :issue:`341`)

Internal changes:

 - Add makefile + dockerfile for simpler testing.
 - Add .gitbugtraq to link comments to issue tracker in GUIs. (:issue:`429`)
 - Add GitHub actions to test PRs and master branch. (:issue:`404`)
 - Remove Travis CI. (:issue:`419`)
 - Remove Appveyor CI and upload coverage report from Windows and Ubuntu from the GitHub actions. (:issue:`455`)
 - Add check if commit is mentioned in the CHANGELOG.rst. (:issue:`457`)
 - Move flake8 config to setup.cfg and add black code formatter. (:issue:`444`)
 - Fix filter/exclude relative path issue in Windows. (:issue:`320`, :issue:`479`) 
 - Extend test framework for CI:

   - Set make variable TEST_OPTS as environment variable inside docker. (:issue:`372`)
   - Add make variable USE_COVERAGE to extend flags for coverage report in GitHub actions. (:issue:`404`)
   - Extend tests to use an unified diff in the assert. Add test options `--generate_reference`,
     `--update_reference` and `--skip_clean`. (:issue:`379`)
   - Support multiple output patterns in integration tests. (:issue:`383`)
   - New option `--archive_differences` to save the different files as ZIP.
     Use this ZIP as artifact in AppVeyor. (:issue:`392`)
   - Add support for gcc-8 to test suite and docker tests. (:issue:`423`)
   - Run as limited user inside docker container and add test with read only directory. (:issue:`445`)

4.2 (6 November 2019)
---------------------

Breaking changes:

 - Dropped support for Python 3.4.
 - Format flag parameters like :option:`--xml` or :option:`--html`
   now take an optional output file name.
   This potentially changes the interpretation of search paths.
   In ``gcovr --xml foo``,
   previous gcovr versions would search the ``foo`` directory for coverage data.
   Now, gcovr will try to write the Cobertura report to the ``foo`` file.
   To keep the old meaning, separate positional arguments like
   ``gcovr --xml -- foo``.

Improvements and new features:

 - :ref:`Configuration file <configuration>` support (experimental).
   (:issue:`167`, :issue:`229`, :issue:`279`, :issue:`281`, :issue:`293`,
   :issue:`300`, :issue:`304`)
 - :ref:`JSON output <json_output>`. (:issue:`301`, :issue:`321`, :issue:`326`)
 - :ref:`Combining tracefiles <combining_tracefiles>`
   with :option:`gcovr --add-tracefile`.
   (:issue:`10`, :issue:`326`)
 - :ref:`SonarQube XML Output <sonarqube_xml_output>`. (:issue:`308`)
 - Handle cyclic symlinks correctly during coverage data search.
   (:issue:`284`)
 - Simplification of :option:`--object-directory` heuristics.
   (:issue:`18`, :issue:`273`, :issue:`280`)
 - Exception-only code like a ``catch`` clause is now shown as uncovered.
   (:issue:`283`)
 - New :option:`--exclude-throw-branches` option
   to exclude exception handler branches. (:issue:`283`)
 - Support ``--root ..`` style invocation,
   which might fix some CMake-related problems. (:issue:`294`)
 - Fix wrong names in report

gcovr/__main__.py
# -*- coding:utf-8 -*-

#  ************************** Copyrights and license ***************************
#
# This file is part of gcovr 4.2, a parsing and reporting tool for gcov.
# https://gcovr.com/en/stable
#
# _____________________________________________________________________________
#
# Copyright (c) 2013-2021 the gcovr authors
# Copyright (c) 2013 Sandia Corporation.
# This software is distributed under the BSD License.
# Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
# the U.S. Government retains certain rights in this software.
# For more information, see the README.rst file.
#
# ****************************************************************************

import os
import re
import sys
import io

from argparse import ArgumentParser
from os.path import normpath
from tempfile import mkdtemp
from shutil import rmtree
from glob import glob

from .configuration import (
    argument_parser_setup, merge_options_and_set_defaults,
    parse_config_file, parse_config_into_dict, OutputOrDefault)
from .gcov import (find_existing_gcov_files, find_datafiles,
                   process_existing_gcov_file, process_datafile)
from .utils import (get_global_stats, AlwaysMatchFilter,
                    DirectoryPrefixFilter, Logger)
from .version import __version__
from .workers import Workers

# generators
from .writer.json import (gcovr_json_files_to_coverage)
from .writer.cobertura import print_xml_report
from .writer.html import print_html_report
from .writer.json import print_json_report, print_json_summary_report
from .writer.txt import print_text_report
from .writer.csv import print_csv_report
from .writer.summary import print_summary
from .writer.sonarqube import print_sonarqube_report
from .writer.coveralls import print_coveralls_report


#
# Exits with status 2 if below threshold
#
def fail_under(covdata, threshold_line, threshold_branch):
    (lines_total, lines_covered, percent,
        #branches_total, branches_covered,
        #percent_branches) = get_global_stats(covdata)
        functions_total, functions_covered, percent_functions,
        branches_total, branches_covered, percent_branches) = get_global_stats(covdata)

    if branches_total == 0:
        percent_branches = 100.0

    if percent < threshold_line and percent_branches < threshold_branch:
        sys.exit(6)
    if percent < threshold_line:
        sys.exit(2)
    if percent_branches < threshold_branch:
        sys.exit(4)


def create_argument_parser():
    """Create the argument parser."""

    parser = ArgumentParser(add_help=False)
    parser.usage = "gcovr [options] [search_paths...]"
    parser.description = \
        "A utility to run gcov and summarize the coverage in simple reports."

    parser.epilog = "See <http://gcovr.com/> for the full manual."

    options = parser.add_argument_group('Options')
    options.add_argument(
        "-h", "--help",
        help="Show this help message, then exit.",
        action="help"
    )
    options.add_argument(
        "--version",
        help="Print the version number, then exit.",
        action="store_true",
        dest="version",
        default=False
    )

    argument_parser_setup(parser, options)

    return parser


COPYRIGHT = (
    "Copyright 2013-2018 the gcovr authors\n"
    "Copyright 2013 Sandia Corporation\n"
    "Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,\n"
    "the U.S. Government retains certain rights in this software."
)


def find_config_name(partial_options):
    cfg_name = getattr(partial_options, 'config', None)
    if cfg_name is not None:
        return cfg_name

    root = getattr(partial_options, 'root', '')
    if root:
        cfg_name = os.path.join(root, 'gcovr.cfg')
    else:
        cfg_name = 'gcovr.cfg'

    if os.path.isfile(cfg_name):
        return cfg_name

    return None


class Options(object):
    def __init__(self, **kwargs):
        self.__dict__.update(kwargs)


def main(args=None):
    parser = create_argument_parser()
    cli_options = parser.parse_args(args=args)

    # load the config
    cfg_name = find_config_name(cli_options)
    cfg_options = {}
    if cfg_name is not None:
        with io.open(cfg_name, encoding='UTF-8') as cfg_file:
            cfg_options = parse_config_into_dict(
                parse_config_file(cfg_file, filename=cfg_name))

    options_dict = merge_options_and_set_defaults(
        [cfg_options, cli_options.__dict__])
    options = Options(**options_dict)

    logger = Logger(options.verbose)

    if cli_options.version:
        logger.msg(
            "gcovr {version}\n"
            "\n"
            "{copyright}",
            version=__version__, copyright=COPYRIGHT)
        sys.exit(0)

    if options.html_title == '':
        logger.error(
            "an empty --html_title= is not allowed.")
        sys.exit(1)

    if options.html_medium_threshold == 0:
        logger.error(
            "value of --html-medium-threshold= should not be zero.")
        sys.exit(1)

    if options.html_medium_threshold > options.html_high_threshold:
        logger.error(
            "value of --html-medium-threshold={} should be\n"
            "lower than or equal to the value of --html-high-threshold={}.",
            options.html_medium_threshold, options.html_high_threshold)
        sys.exit(1)

    if options.html_tab_size < 1:
        logger.error(
            "value of --html-tab-size= should be greater 0.")
        sys.exit(1)

    potential_html_output = (
        (options.html and options.html.value)
        or (options.html_details and options.html_details.value)
        or (options.output and options.output.value))
    if options.html_details and not potential_html_output:
        logger.error(
            "a named output must be given, if the option --html-details\n"
            "is used.")
        sys.exit(1)

    if options.html_self_contained is False and not potential_html_output:
        logger.error(
            "can only disable --html-self-contained when a named output is given.")
        sys.exit(1)

    if options.objdir is not None:
        if not options.objdir:
            logger.error(
                "empty --object-directory option.\n"
                "\tThis option specifies the path to the object file "
                "directory of your project.\n"
                "\tThis option cannot be an empty string.")
            sys.exit(1)
        tmp = options.objdir.replace('/', os.sep).replace('\\', os.sep)
        while os.sep + os.sep in tmp:
            tmp = tmp.replace(os.sep + os.sep, os.sep)
        if normpath(options.objdir) != tmp:
            logger.warn(
                "relative referencing in --object-directory.\n"
                "\tthis could cause strange errors when gcovr attempts to\n"
                "\tidentify the original gcc working directory.")
        if not os.path.exists(normpath(options.objdir)):
            logger.error(
                "Bad --object-directory option.\n"
                "\tThe specified directory does not exist.")
            sys.exit(1)

    options.starting_dir = os.path.abspath(os.getcwd())
    if not options.root:
        logger.error(
            "empty --root option.\n"
            "\tRoot specifies the path to the root "
            "directory of your project.\n"
            "\tThis option cannot be an empty string.")
        sys.exit(1)
    options.root_dir = os.path.abspath(options.root)

    #
    # Setup filters
    #

    # The root filter isn't technically a filter,
    # but is used to turn absolute paths into relative paths
    options.root_filter = re.compile('^' + re.escape(options.root_dir + os.sep))

    if options.exclude_dirs is not None:
        options.exclude_dirs = [
            f.build_filter(logger) for f in options.exclude_dirs]

    options.exclude = [f.build_filter(logger) for f in options.exclude]
    options.filter = [f.build_filter(logger) for f in options.filter]
    if not options.filter:
        options.filter = [DirectoryPrefixFilter(options.root_dir)]

    options.gcov_exclude = [
        f.build_filter(logger) for f in options.gcov_exclude]
    options.gcov_filter = [f.build_filter(logger) for f in options.gcov_filter]
    if not options.gcov_filter:
        options.gcov_filter = [AlwaysMatchFilter()]

    # Output the filters for debugging
    for name, filters in [
        ('--root', [options.root_filter]),
        ('--filter', options.filter),
        ('--exclude', options.exclude),
        ('--gcov-filter', options.gcov_filter),
        ('--gcov-exclude', options.gcov_exclude),
        ('--exclude-directories', options.exclude_dirs),
    ]:
        logger.verbose_msg('Filters for {}: ({})', name, len(filters))
        for f in filters:
            logger.verbose_msg('- {}', f)

    if options.exclude_lines_by_pattern:
        try:
            re.compile(options.exclude_lines_by_pattern)
        except re.error as e:
            logger.error(
                "--exclude-lines-by-pattern: "
                "Invalid regular expression: {}, error: {}",
                repr(options.exclude_lines_by_pattern), e)
            sys.exit(1)

    covdata = dict()
    if options.add_tracefile:
        collect_coverage_from_tracefiles(covdata, options, logger)
    else:
        collect_coverage_from_gcov(covdata, options, logger)

    logger.verbose_msg("Gathered coveraged data for {} files", len(covdata))

    # Print reports
    error_occurred = print_reports(covdata, options, logger)
    if error_occurred:
        logger.error(
            "Error occurred while printing reports"
        )
        sys.exit(7)

    if options.fail_under_line > 0.0 or options.fail_under_branch > 0.0:
        fail_under(covdata, options.fail_under_line, options.fail_under_branch)


def collect_coverage_from_tracefiles(covdata, options, logger):
    datafiles = set()

    for trace_files_regex in options.add_tracefile:
        trace_files = glob(trace_files_regex, recursive=True)
        if not trace_files:
            logger.error(
                "Bad --add-tracefile option.\n"
                "\tThe specified file does not exist.")
            sys.exit(1)
        else:
            for trace_file in trace_files:
                datafiles.add(normpath(trace_file))

    options.root_dir = os.path.abspath(options.root)
    gcovr_json_files_to_coverage(datafiles, covdata, options)


def collect_coverage_from_gcov(covdata, options, logger):
    datafiles = set()

    find_files = find_datafiles
    process_file = process_datafile
    if options.gcov_files:
        find_files = find_existing_gcov_files
        process_file = process_existing_gcov_file

    # Get data files
    if not options.search_paths:
        options.search_paths = [options.root]

        if options.objdir is not None:
            options.search_paths.append(options.objdir)

    for search_path in options.search_paths:
        datafiles.update(find_files(search_path, logger, options.exclude_dirs))

    # Get coverage data
    with Workers(options.gcov_parallel, lambda: {
                 'covdata': dict(),
                 'workdir': mkdtemp(),
                 'toerase': set(),
                 'options': options}) as pool:
        logger.verbose_msg("Pool started with {} threads", pool.size())
        for file_ in datafiles:
            pool.add(process_file, file_)
        contexts = pool.wait()

    toerase = set()
    for context in contexts:
        for fname, cov in context['covdata'].items():
            if fname not in covdata:
                covdata[fname] = cov
            else:
                covdata[fname].update(cov)
        toerase.update(context['toerase'])
        rmtree(context['workdir'])
    for filepath in toerase:
        if os.path.exists(filepath):
            os.remove(filepath)


def print_reports(covdata, options, logger):
    generators = []

    if options.txt:
        generators.append((
            [options.txt],
            print_text_report,
            lambda: logger.warn(
                "Text output skipped - "
                "consider providing an output file with `--txt=OUTPUT`.")))

    if options.xml or options.prettyxml:
        generators.append((
            [options.xml],
            print_xml_report,
            lambda: logger.warn(
                "Cobertura output skipped - "
                "consider providing an output file with `--xml=OUTPUT`.")))

    if options.html or options.html_details:
        generators.append((
            [options.html, options.html_details],
            print_html_report,
            lambda: logger.warn(
                "HTML output skipped - "
                "consider providing an output file with `--html=OUTPUT`.")))

    if options.sonarqube:
        generators.append((
            [options.sonarqube],
            print_sonarqube_report,
            lambda: logger.warn(
                "Sonarqube output skipped - "
                "consider providing output file with `--sonarqube=OUTPUT`.")))

    if options.json or options.json_pretty:
        generators.append((
            [options.json],
            print_json_report,
            lambda: logger.warn(
                "JSON output skipped - "
                "consider providing output file with `--json=OUTPUT`.")))

    if options.json_summary or options.json_summary_pretty:
        generators.append((
            [options.json_summary],
            print_json_summary_report,
            lambda: logger.warn(
                "JSON summary output skipped - "
                "consider providing output file with `--json-summary=OUTPUT`.")))

    if options.csv:
        generators.append((
            [options.csv],
            print_csv_report,
            lambda: logger.warn(
                "CSV output skipped - "
                "consider providing output file with `--csv=OUTPUT`.")))

    if options.coveralls or options.coveralls_pretty:
        generators.append((
            [options.coveralls],
            print_coveralls_report,
            lambda: logger.warn(
                "Coveralls output skipped - "
                "consider providing output file with `--coveralls=OUTPUT`.")))

    generator_error_occurred = False
    reports_were_written = False
    default_output_used = False
    default_output = OutputOrDefault(None) if options.output is None else options.output

    for output_choices, generator, on_no_output in generators:
        output = OutputOrDefault.choose(output_choices, default=default_output)
        if output is not None and output is default_output:
            default_output_used = True
            if not output.is_dir:
                default_output = None
        if output is not None:
            if generator(covdata, output.abspath, options):
                generator_error_occurred = True
            reports_were_written = True
        else:
            on_no_output()

    if not reports_were_written:
        print_text_report(covdata, '-' if default_output is None else default_output.abspath, options)
        default_output = None

    if default_output is not None and default_output.value is not None and not default_output_used:
        logger.warn("--output={!r} option was provided but not used.",
                    default_output.value)

    if options.print_summary:
        print_summary(covdata)

    return generator_error_occurred


if __name__ == '__main__':
    main()
